### PR TITLE
[torch-frontend]add canonicalize for ccl.collectives

### PR DESCRIPTION
**Pull request by linuxlonelyeagle**

### CODE REVIEW

File: compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp [hunk 1]
```diff
@@ -170,6 +171,65 @@ LogicalResult BroadcastOp::verify() {
...
+      op.setSynchronous(true);
+      rewriter.replaceOp(nextOp, op);
+      return success();
+    }
+    return failure();
+  }
+};
+
+// Eliminate the WaitOp after a BroadcastOp which synchronous is true.
+struct EliminateUnnecessaryWait : public OpRewritePattern<BroadcastOp> {
```
Comment (qingyunqu) [inline compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp:209]: add lit test?

**Review by qingyunqu — CHANGES_REQUESTED** [review]

File: compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp [hunk 2]
```diff
@@ -76,6 +77,21 @@ LogicalResult verifyP2PIndex(std::optional<Location> location,
...
 }
 
+// A BroadcastOp with a synchronous of false plus a WaitOp equals a BroadcastOp
+// with a synchronous of true.
+template <typename T> struct CombineCclOpAndWait : public OpRewritePattern<T> {
+  using OpRewritePattern<T>::OpRewritePattern;
+  LogicalResult matchAndRewrite(T op,
+                                PatternRewriter &rewriter) const override {
+    for (auto user : op.getResult().getUsers()) {
+      if (isa<ccl::WaitOp>(user)) {
```
Comment (Vremold) [inline compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp:87]: Please use stricter checks to require all users to be ccl.wait

File: compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp [hunk 3]
```diff
@@ -76,6 +77,21 @@ LogicalResult verifyP2PIndex(std::optional<Location> location,
   return success();
 }
 
+// A BroadcastOp with a synchronous of false plus a WaitOp equals a BroadcastOp
```
Comment (Vremold) [inline compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp:80]: Refine this comment as this pattern is not only for broadcast op

File: compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp [hunk 4]
```diff
@@ -93,6 +109,33 @@ void CclDialect::initialize() {
...
 //===----------------------------------------------------------------------===//
 
+namespace {
+
+// Eliminate the WaitOp after a BroadcastOp which synchronous is true.
+struct EliminateUnnecessaryWait : public OpRewritePattern<WaitOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(WaitOp op,
+                                PatternRewriter &rewriter) const override {
+    auto cclOp = op.getSrc().getDefiningOp();
```
Comment (Vremold) [inline compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp:119]: Please check whether its src is really a ccl.op.

**Review by Vremold — CHANGES_REQUESTED** [review]
LGTM overall. Just a few comments.

**Review by qingyunqu — APPROVED** [review]
LGTM

**Pull request merged**